src/commands/SinceCommand.ts

```diff
@@ -0,0 +1,275 @@
+    const [dateOrDurationToken, actionToken, maxEntriesToken, ...optionalTokens] = tokens;
+
+    // Parse origin date or duration.
+    const minDateResult = parseToken("<date>/<duration>", dateOrDurationToken, source => {
+        // Attempt to parse `<date>/<duration>` as a date.
+        let maybeMinDate = new Date(source);
+        let maybeMaxAgeMS = Date.now() - maybeMinDate.getTime() as number;
+        if (!Number.isNaN(maybeMaxAgeMS)) {
+            return { ok: { minDate: maybeMinDate, maxAgeMS: maybeMaxAgeMS} };
+        }
+
+        //...or as a duration
+        maybeMaxAgeMS = parseDuration(source);
+        if (maybeMaxAgeMS && !Number.isNaN(maybeMaxAgeMS)) {
+            maybeMaxAgeMS = Math.abs(maybeMaxAgeMS);
+            return { ok: { minDate: new Date(Date.now() - maybeMaxAgeMS), maxAgeMS: maybeMaxAgeMS } }
+        }
+        return { error: "" };
+    });
+    if ("error" in minDateResult) {
+        return minDateResult;
+    }
+    const { minDate, maxAgeMS } = minDateResult.ok!;
+
+    // Parse max entries.
+    const maxEntriesResult = parseToken("<maxEntries>", maxEntriesToken, source => {
+        const maybeMaxEntries = Number.parseInt(source, 10);
+        if (Number.isNaN(maybeMaxEntries)) {
+            return { error: "Not a number" };
+        } else {
+            return { ok: maybeMaxEntries };
+        }
+    });
+    if ("error" in maxEntriesResult) {
+        return maxEntriesResult;
+    }
+    const maxEntries = maxEntriesResult.ok!;
+
+    // Attempt to parse `<action>` as Action.
+    const actionResult = parseToken("<action>", actionToken, source => {
+        for (let key in Action) {
+            const maybeAction = Action[key as keyof typeof Action];
+            if (key === source) {
+                return { ok: maybeAction }
+            } else if (maybeAction === source) {
+                return { ok: maybeAction }
+            }
+        }
+        return {error: `Expected one of ${JSON.stringify(Action)}`};
+    })
+    if ("error" in actionResult) {
+        return actionResult;
+    }
+    const action: Action = actionResult.ok!;
+
+    // Now list affected rooms.
+    const rooms: Set<string> = new Set();
+    let reasonParts: string[] | undefined;
+    for (let token of optionalTokens) {
+        const maybeArg = getTokenAsString(reasonParts ? "[reason]" : "[room]", token);
+        if ("error" in maybeArg) {
+            return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", maybeArg.error);
+        }
+        const maybeRoom = maybeArg.ok;
+        if (!reasonParts) {
+            // If we haven't reached the reason yet, attempt to use `maybeRoom` as a room.
+            if (maybeRoom === "*") {
+                for (let roomId of Object.keys(mjolnir.protectedRooms)) {
+                    rooms.add(roomId);
+                }
+                continue;
+            } else if (maybeRoom.startsWith("#") || maybeRoom.startsWith("!")) {
```

Comment:
![image](https://user-images.githubusercontent.com/50846879/157243446-63d4d1bb-f478-4040-8629-abbe4d56d186.png)

are we sure this works well with aliases/pills?